sql/src/main/java/io/crate/planner/consumer/MultiSourceGroupByConsumer.java

```diff
@@ -0,0 +1,231 @@
+        return visitor.process(relation, context);
+    }
+
+    private static final JoinConditionFieldCollector FIELD_COLLECTOR = new JoinConditionFieldCollector();
+
+    private static class JoinConditionFieldCollector extends DefaultTraversalSymbolVisitor<List<Field>, Void> {
+        @Override
+        public Void visitField(Field symbol, List<Field> context) {
+            context.add(symbol);
+            return null;
+        }
+    }
+
+    private static class Visitor extends RelationPlanningVisitor {
+
+        private final Functions functions;
+
+        public Visitor(Functions functions) {
+            this.functions = functions;
+        }
+
+        @Override
+        public Plan visitMultiSourceSelect(MultiSourceSelect multiSourceSelect, ConsumerContext context) {
+            QuerySpec querySpec = multiSourceSelect.querySpec();
+
+            // Check if group by is present - if not continue with next consumer in chain.
+            if (!querySpec.groupBy().isPresent()) {
+                return null;
+            }
+
+            // Copy because MSS planning mutates symbols.
+            querySpec = querySpec.copyAndReplace(i -> i);
+
+            List<Symbol> groupKeys = querySpec.groupBy().get();
+            SplitPoints splitPoints = SplitPoints.create(querySpec);
+
+            List<Symbol> outputs = querySpec.outputs();
+            if (querySpec.hasAggregates() == true) {
+                querySpec.hasAggregates(false);
+                querySpec.outputs(splitPoints.toCollect());
+                // TODO: Alter the fields on multiSourceSelect.
+            }
+
+            this.removePostGroupingActionsFromQuerySpec(multiSourceSelect, splitPoints);
+            Plan plan = this.createExecutionPlan(context, multiSourceSelect);
+            this.addGroupProtection(plan, splitPoints, groupKeys);
+            this.addFilterProtection(plan, groupKeys, splitPoints, querySpec);
+            this.addPostAggregationProjection(plan, context, splitPoints, groupKeys, querySpec, outputs);
+
+            return plan;
+        };
+
+        /**
+         * Create and return the current execution plan.
+         */
+        private static Plan createExecutionPlan(ConsumerContext context, MultiSourceSelect multiSourceSelect) {
+            Plan plan = context.plannerContext().planSubRelation(multiSourceSelect, context);
+            boolean executesOnHandler = ExecutionPhases.executesOnHandler(
+                context.plannerContext().handlerNode(),
+                plan.resultDescription().nodeIds()
+            );
+            if (!executesOnHandler) {
+                // TODO: support distributed execution.
+                throw new UnsupportedFeatureException("Distribution execution is not supported");
+            }
+
+            return plan;
+        }
+
+        /**
+         * Adds the post aggregation protection to the given plan.
+         */
+        private static void addPostAggregationProjection(
```

Comment:
would rename this to `addTopN` maybe with a comment that this also re-orders column after groupBy messed them up.
I think in the non-grouping aggregate consumer the postAggregationProjection naming is used because there it is a list which contains the other projections as well